src/Middleware/AuthTokenMiddleware.php

```diff
@@ -99,7 +105,20 @@ public function __invoke(callable $handler)
                 return $handler($request, $options);
             }
 
-            $request = $request->withHeader('authorization', 'Bearer ' . $this->fetchToken());
+            if (!$this->fetcher instanceof UpdateMetadataInterface ||
+                ($this->fetcher instanceof FetchAuthTokenCache && !$this->fetcher->getFetcher() instanceof UpdateMetadataInterface)
+            ) {
+                $request =$request->withHeader('authorization', 'Bearer ' . $this->fetchToken());
+            } else {
+                $headers = $this->fetcher->updateMetadata($request->getHeaders(), null, $this->httpHandler);
+                $request = Utils::modifyRequest($request, ['set_headers' => $headers]);
+                if ($this->tokenCallback) {
+                    $auth_tokens = $this->fetcher->fetchAuthToken($this->httpHandler);
+                    if (array_key_exists('access_token', $auth_tokens)) {
+                        call_user_func($this->tokenCallback, $this->fetcher->getCacheKey(), $auth_tokens['access_token']);
+                    }
+                }
+            }
```

Comment:
A few things here: 
1. This will call `fetchAuthToken` twice, which we don't want (it would end up making multiple HTTP requests in some cases). 
2. This behavior (using `tokenCallback`) is essentially deprecated (we do not want to support it in auth V2), but since we'd like to maintain BC, we can get the access token from the auth header
3. Since this is being applied to ALL calls in the current model, it should be outside the `else` statement to ensure that BC doesn't break
```suggestion
            } else {
                $headers = $this->fetcher->updateMetadata($request->getHeaders(), null, $this->httpHandler);
                $request = Utils::modifyRequest($request, ['set_headers' => $headers]);
            }

            // support tokenCallback for compatibility (this should be removed in the next major version)
            if ($this->tokenCallback && $bearer = $request->getHeaderLine('authorization')) {
                if (0 === strpos($bearer, 'Bearer ')) {
                    call_user_func($this->tokenCallback, $this->fetcher->getCacheKey(), substr($bearer, 7));
                }
           }

```

Reply:
This approach has two problems it seems:

1) We need to call token callback only in the case of  an `access_token` which seems impossible to determine from `updateMetadata()`'s response. 
    1) `lastTokenReceived` method solved this problem but it seems we are deprecating it.
2) Current implementation doesn't retries the `updateMetadata` call, we need to change it I think.

Reply:
> `fetchToken` retries the fetching calls using exponential backoff.

I am confused by this - retries are not represented in `fetchToken`, and exponential backoff doesn't exist in the auth library.

> We need to call token callback only in the case of an `access_token`

Ahh good call, I missed that this only happened previously with `access_token` and not with `id_token`. 

> `lastTokenReceived` method solved this problem but it seems we are deprecating it.

True, and I agree that in this case (to preserve BC, and to not call `fetchAuthToken` twice) it's the best solution to use `getLastReceivedToken`. Yes we want to discourage the use of `getLastReceivedToken`, but we  _also_ want to discourage using `tokenCallback` (both these "features" are technical debt that should be removed). 

... So changing my code above, something like this should work instead: 

```php

if ($this->tokenCallback && $token = $this->fetcher->getLastReceivedToken()) {
    if (isset($token['access_token'])) {
        call_user_func($this->tokenCallback, $this->fetcher->getCacheKey(), $token['access_token']);
    }
}
```

...And then we can remove the `tokenCallback` logic from `fetchToken`, and keep it all in one place

Reply:
> I am confused by this - retries are not represented in fetchToken, and exponential backoff doesn't exist in the auth library.

I’m sorry I created a confusion by saying stuff about `ExponentialBackoff`. At that moment, somehow I mixed google-cloud-php’s Core library knowledge with this.  

> True, and I agree that in this case (to preserve BC, and to not call fetchAuthToken twice) it's the best solution to use getLastReceivedToken. Yes we want to discourage the use of getLastReceivedToken, but we also want to discourage using tokenCallback (both these "features" are technical debt that should be removed).

Yess! I was thinking along the same lines that we should go ahead with using `getlastFetchedToken` method in the case of invoking `tokenCallback`.